File: RemoveDuplicatesFromSortedList.java
/**
 * Author: Chih-Jye Wang
 * Date  : Feb 12, 2015
 *
 * Given a sorted linked list, delete all duplicates such that each element appear only once.
 *
 * For example,
 * Given 1->1->2, return 1->2.
 * Given 1->1->2->3->3, return 1->2->3.
 */
public class RemoveDuplicatesFromSortedList {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode front = head;
        ListNode rear  = head;
        while(front != null) {
            if(rear == null || front.val != rear.val) {
                front.next = rear;
                front = rear;    
            }
            else
                rear = rear.next;
        }
        return head;
    }
}

File: ruby/remove_duplicates_from_sorted_list.rb
# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val)
        @val = val
        @next = nil
    end
end

# @param {ListNode} head
# @return {ListNode}
def delete_duplicates(head)
    i = head
    while i
        break if !i.next
        while i.next && i.next.val == i.val
            i.next = i.next.next
        end
        i = i.next
    end
    return head
end
